web_poet/fields.py

```diff
@@ -200,19 +220,43 @@ async def item_from_fields(
 
 
 def item_from_fields_sync(
-    obj, item_cls: Type[T] = dict, *, skip_nonitem_fields: bool = False  # type: ignore[assignment]
+    obj,
+    item_cls: Type[T] = dict,  # type: ignore[assignment]
+    *,
+    skip_nonitem_fields: bool = False,
+    field_names: Optional[Iterable[str]] = None,
 ) -> T:
     """Synchronous version of :func:`item_from_fields`."""
-    field_names = list(get_fields_dict(obj))
+    if field_names is None:
+        field_names = list(get_fields_dict(obj))
     if skip_nonitem_fields:
         field_names = _without_unsupported_field_names(item_cls, field_names)
     return item_cls(**{name: getattr(obj, name) for name in field_names})
 
 
 def _without_unsupported_field_names(
-    item_cls: type, field_names: List[str]
+    item_cls: type, field_names: Iterable[str]
 ) -> List[str]:
     item_field_names = ItemAdapter.get_field_names_from_class(item_cls)
     if item_field_names is None:  # item_cls doesn't define field names upfront
-        return field_names[:]
+        return list(field_names)
     return list(set(field_names) & set(item_field_names))
+
+
+@attrs.define
+class SelectFields:
+    """This is used as a dependency in a page object to control which fields it
+    would populate the item class that it returns.
+    """
+
+    #: Fields that the page object would use to populate its item class. It's a
+    #: mapping of field names to boolean values to where ``True`` would indicate
+    #: it being included in ``.to_item()`` calls.
+    fields: Optional[MutableMapping[str, bool]] = None
+
+    #: Controls what happens when encountering an unknown field. For example,
+    #: an unknown field was passed and the page object doesn't recognize it.
+    #:
+    #: Setting it to ``"raise"`` would raise an :class:`AttributeError`,
+    #: ``"warn"`` produces a :class:`UserWarning`, while ``"ignore"`` does nothing.
+    on_unknown_field: UnknownFieldActions = "raise"
```

Comment:
Hey! Could you please elaborate on the use cases for on_unknown_field?

Also, what does unknown field mean? Does it mean that the field is not in the item, or does it mean that the field is not defined on the page object?

Reply:
> does it mean that the field is not defined on the page object?

It's mostly regarding about this but because of the parity between the page object and its item class, the field should generally not be in the item as well.

The unknown field can be derived by something like `set(fields_to_include) - set(fields_to_exclude)`. This means unknown fields that are to be excluded won't be considered as actual unknown fields.

I'm also thinking of removing this parameter and simply raise any unknown fields by default. What do you think?

Reply:
Aha, I see. 

It seems the case with field not in the item and a field not in the page object are very different. 

If a field is not in the item, we should probably raise an exception. But there is a case where skip_nonitem_fields=True is used; in this case it actually could make sense to have fields in SelectFields which are not in the item.

If a field is not defined in the page object, but is present in the item, then final item.field returned by to_item is going to have a default value, None in most cases. It could also be the case that the attribute is populated in to_item method without using web-poet fields. It's not an error. It seems that we should allow to pick any fields which are present in the item, even if they're not fields of the page object.

We should have a clear answer though what happens when some of the item attributes are extracted using `@field`s, and some of them are added later, in the to_item method.

There is also a complicated case where there is a field defined, but then the value assigned to the final item is changed in to_item method. Probably it's fine not to support this case, but we should at least document it.

Overall the problem looks more complicated than I initially thought :)

Reply:
When I suggested such an option, I was thinking of a scenario like this one:

B team writes `BPage`, subclassing `APage` from A team, and returning `BItem`, which does *not* subclass `AItem`. B team might not want new fields in `APage` to break `BPage`. But they might want a warning or error to look into it eventually in case we want to do something about it.

But it can indeed be more complex than that. When a field is removed, B team may want a different resolution. For example, they might want a warning for new fields, but an exception for missing fields.